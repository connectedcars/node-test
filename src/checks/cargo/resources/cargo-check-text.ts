import { CargoMessage } from '../cargo-types'

export const cargoCheckFailedOutput: CargoMessage[] = [
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused import: `std::io::BufReader`\n --> src/main.rs:3:5\n  |\n3 | use std::io::BufReader;\n  |     ^^^^^^^^^^^^^^^^^^\n  |\n  = note: `#[warn(unused_imports)]` on by default\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_imports)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'remove the whole `use` item',
          rendered: null,
          spans: [
            {
              byte_end: 52,
              byte_start: 29,
              column_end: 24,
              column_start: 1,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 3,
              line_start: 3,
              suggested_replacement: '',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 24,
                  highlight_start: 1,
                  text: 'use std::io::BufReader;'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_imports',
        explanation: null
      },
      level: 'warning',
      message: 'unused import: `std::io::BufReader`',
      spans: [
        {
          byte_end: 51,
          byte_start: 33,
          column_end: 23,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 3,
          line_start: 3,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 23,
              highlight_start: 5,
              text: 'use std::io::BufReader;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused import: `std::io::BufReader`\n --> src/main.rs:3:5\n  |\n3 | use std::io::BufReader;\n  |     ^^^^^^^^^^^^^^^^^^\n  |\n  = note: `#[warn(unused_imports)]` on by default\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_imports)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'remove the whole `use` item',
          rendered: null,
          spans: [
            {
              byte_end: 52,
              byte_start: 29,
              column_end: 24,
              column_start: 1,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 3,
              line_start: 3,
              suggested_replacement: '',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 24,
                  highlight_start: 1,
                  text: 'use std::io::BufReader;'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_imports',
        explanation: null
      },
      level: 'warning',
      message: 'unused import: `std::io::BufReader`',
      spans: [
        {
          byte_end: 51,
          byte_start: 33,
          column_end: 23,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 3,
          line_start: 3,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 23,
              highlight_start: 5,
              text: 'use std::io::BufReader;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused variable: `foo`\n  --> src/main.rs:15:11\n   |\n15 |       let foo = foo();\n   |           ^^^ help: if this is intentional, prefix it with an underscore: `_foo`\n   |\n   = note: `#[warn(unused_variables)]` on by default\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_variables)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'if this is intentional, prefix it with an underscore',
          rendered: null,
          spans: [
            {
              byte_end: 160,
              byte_start: 157,
              column_end: 14,
              column_start: 11,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 15,
              line_start: 15,
              suggested_replacement: '_foo',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 14,
                  highlight_start: 11,
                  text: '      let foo = foo();'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_variables',
        explanation: null
      },
      level: 'warning',
      message: 'unused variable: `foo`',
      spans: [
        {
          byte_end: 160,
          byte_start: 157,
          column_end: 14,
          column_start: 11,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 15,
          line_start: 15,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 14,
              highlight_start: 11,
              text: '      let foo = foo();'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused variable: `foo`\n  --> src/main.rs:15:11\n   |\n15 |       let foo = foo();\n   |           ^^^ help: if this is intentional, prefix it with an underscore: `_foo`\n   |\n   = note: `#[warn(unused_variables)]` on by default\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_variables)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'if this is intentional, prefix it with an underscore',
          rendered: null,
          spans: [
            {
              byte_end: 160,
              byte_start: 157,
              column_end: 14,
              column_start: 11,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 15,
              line_start: 15,
              suggested_replacement: '_foo',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 14,
                  highlight_start: 11,
                  text: '      let foo = foo();'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_variables',
        explanation: null
      },
      level: 'warning',
      message: 'unused variable: `foo`',
      spans: [
        {
          byte_end: 160,
          byte_start: 157,
          column_end: 14,
          column_start: 11,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 15,
          line_start: 15,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 14,
              highlight_start: 11,
              text: '      let foo = foo();'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused variable: `f`\n  --> src/main.rs:20:9\n   |\n20 |     let f = File::open("foo.txt");\n   |         ^ help: if this is intentional, prefix it with an underscore: `_f`\n\n',
      children: [
        {
          code: null,
          level: 'help',
          message: 'if this is intentional, prefix it with an underscore',
          rendered: null,
          spans: [
            {
              byte_end: 213,
              byte_start: 212,
              column_end: 10,
              column_start: 9,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 20,
              line_start: 20,
              suggested_replacement: '_f',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 10,
                  highlight_start: 9,
                  text: '    let f = File::open("foo.txt");'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_variables',
        explanation: null
      },
      level: 'warning',
      message: 'unused variable: `f`',
      spans: [
        {
          byte_end: 213,
          byte_start: 212,
          column_end: 10,
          column_start: 9,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 20,
          line_start: 20,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 10,
              highlight_start: 9,
              text: '    let f = File::open("foo.txt");'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused variable: `f`\n  --> src/main.rs:20:9\n   |\n20 |     let f = File::open("foo.txt");\n   |         ^ help: if this is intentional, prefix it with an underscore: `_f`\n\n',
      children: [
        {
          code: null,
          level: 'help',
          message: 'if this is intentional, prefix it with an underscore',
          rendered: null,
          spans: [
            {
              byte_end: 213,
              byte_start: 212,
              column_end: 10,
              column_start: 9,
              expansion: null,
              file_name: 'src/main.rs',
              is_primary: true,
              label: null,
              line_end: 20,
              line_start: 20,
              suggested_replacement: '_f',
              suggestion_applicability: 'MachineApplicable',
              text: [
                {
                  highlight_end: 10,
                  highlight_start: 9,
                  text: '    let f = File::open("foo.txt");'
                }
              ]
            }
          ]
        }
      ],
      code: {
        code: 'unused_variables',
        explanation: null
      },
      level: 'warning',
      message: 'unused variable: `f`',
      spans: [
        {
          byte_end: 213,
          byte_start: 212,
          column_end: 10,
          column_start: 9,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 20,
          line_start: 20,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 10,
              highlight_start: 9,
              text: '    let f = File::open("foo.txt");'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: value assigned to `foo` is never read\n  --> src/main.rs:29:9\n   |\n29 |     let mut foo = 0;\n   |         ^^^^^^^\n   |\n   = note: `#[warn(unused_assignments)]` on by default\n   = help: maybe it is overwritten before being read?\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_assignments)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'maybe it is overwritten before being read?',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_assignments',
        explanation: null
      },
      level: 'warning',
      message: 'value assigned to `foo` is never read',
      spans: [
        {
          byte_end: 355,
          byte_start: 348,
          column_end: 16,
          column_start: 9,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 29,
          line_start: 29,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 16,
              highlight_start: 9,
              text: '    let mut foo = 0;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: value assigned to `foo` is never read\n  --> src/main.rs:29:9\n   |\n29 |     let mut foo = 0;\n   |         ^^^^^^^\n   |\n   = note: `#[warn(unused_assignments)]` on by default\n   = help: maybe it is overwritten before being read?\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_assignments)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'help',
          message: 'maybe it is overwritten before being read?',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_assignments',
        explanation: null
      },
      level: 'warning',
      message: 'value assigned to `foo` is never read',
      spans: [
        {
          byte_end: 355,
          byte_start: 348,
          column_end: 16,
          column_start: 9,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 29,
          line_start: 29,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 16,
              highlight_start: 9,
              text: '    let mut foo = 0;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: value assigned to `bar` is never read\n  --> src/main.rs:33:5\n   |\n33 |     bar = foo;\n   |     ^^^\n   |\n   = help: maybe it is overwritten before being read?\n\n',
      children: [
        {
          code: null,
          level: 'help',
          message: 'maybe it is overwritten before being read?',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_assignments',
        explanation: null
      },
      level: 'warning',
      message: 'value assigned to `bar` is never read',
      spans: [
        {
          byte_end: 405,
          byte_start: 402,
          column_end: 8,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 33,
          line_start: 33,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 8,
              highlight_start: 5,
              text: '    bar = foo;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: value assigned to `bar` is never read\n  --> src/main.rs:33:5\n   |\n33 |     bar = foo;\n   |     ^^^\n   |\n   = help: maybe it is overwritten before being read?\n\n',
      children: [
        {
          code: null,
          level: 'help',
          message: 'maybe it is overwritten before being read?',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_assignments',
        explanation: null
      },
      level: 'warning',
      message: 'value assigned to `bar` is never read',
      spans: [
        {
          byte_end: 405,
          byte_start: 402,
          column_end: 8,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 33,
          line_start: 33,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 8,
              highlight_start: 5,
              text: '    bar = foo;'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused `std::result::Result` that must be used\n  --> src/main.rs:25:5\n   |\n25 |     f.read(&mut buf);\n   |     ^^^^^^^^^^^^^^^^^\n   |\n   = note: `#[warn(unused_must_use)]` on by default\n   = note: this `Result` may be an `Err` variant, which should be handled\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_must_use)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'note',
          message: 'this `Result` may be an `Err` variant, which should be handled',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_must_use',
        explanation: null
      },
      level: 'warning',
      message: 'unused `std::result::Result` that must be used',
      spans: [
        {
          byte_end: 336,
          byte_start: 319,
          column_end: 22,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 25,
          line_start: 25,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 22,
              highlight_start: 5,
              text: '    f.read(&mut buf);'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered:
        'warning: unused `std::result::Result` that must be used\n  --> src/main.rs:25:5\n   |\n25 |     f.read(&mut buf);\n   |     ^^^^^^^^^^^^^^^^^\n   |\n   = note: `#[warn(unused_must_use)]` on by default\n   = note: this `Result` may be an `Err` variant, which should be handled\n\n',
      children: [
        {
          code: null,
          level: 'note',
          message: '`#[warn(unused_must_use)]` on by default',
          rendered: null,
          spans: []
        },
        {
          code: null,
          level: 'note',
          message: 'this `Result` may be an `Err` variant, which should be handled',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'unused_must_use',
        explanation: null
      },
      level: 'warning',
      message: 'unused `std::result::Result` that must be used',
      spans: [
        {
          byte_end: 336,
          byte_start: 319,
          column_end: 22,
          column_start: 5,
          expansion: null,
          file_name: 'src/main.rs',
          is_primary: true,
          label: null,
          line_end: 25,
          line_start: 25,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 22,
              highlight_start: 5,
              text: '    f.read(&mut buf);'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered: 'warning: 6 warnings emitted\n\n',
      children: [],
      code: null,
      level: 'warning',
      message: '6 warnings emitted',
      spans: []
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    message: {
      rendered: 'warning: 6 warnings emitted\n\n',
      children: [],
      code: null,
      level: 'warning',
      message: '6 warnings emitted',
      spans: []
    }
  },
  {
    reason: 'compiler-artifact',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    profile: {
      opt_level: '0',
      debuginfo: 2,
      debug_assertions: true,
      overflow_checks: true,
      test: false
    },
    features: [],
    filenames: ['/checks-test/target/debug/deps/libcheck_test-3e8dea89f4957cbe.rmeta'],
    executable: null,
    fresh: false
  },
  {
    reason: 'compiler-artifact',
    package_id: 'checks-test 0.1.0 (path+file:///checks-test)',
    target: {
      kind: ['bin'],
      crate_types: ['bin'],
      name: 'checks-test',
      src_path: '/checks-test/src/main.rs',
      edition: '2018',
      doctest: false,
      test: true
    },
    profile: {
      opt_level: '0',
      debuginfo: 2,
      debug_assertions: true,
      overflow_checks: true,
      test: true
    },
    features: [],
    filenames: ['/checks-test/target/debug/deps/libcheck_test-b71ac6c54ee2a77f.rmeta'],
    executable: null,
    fresh: false
  },
  {
    reason: 'build-finished',
    success: true
  }
]
