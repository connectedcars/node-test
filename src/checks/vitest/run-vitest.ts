import { runJsonCommand } from '../checks-common'
import { FormattedTestResults } from './vitest'

export async function runVitest(command = 'vitest', extraArgs: string[] = []): Promise<FormattedTestResults> {
  const [, json] = await runJsonCommand<FormattedTestResults>(
    command,
    [...extraArgs, '--run', '--silent', '--no-color', '--reporter=json'],
    {
      env: { ...process.env, TZ: 'UTC' }
    }
  )
  return json
}
