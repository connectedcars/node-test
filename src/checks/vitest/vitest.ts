import { CheckAnnotation, CheckAnnotationLevel, CheckRunCompleted } from '../checks-common'

// Types copied from vitest:
// https://github.com/vitest-dev/vitest/blob/445a7b6d4ab906c05c211e7b1a8102779fa35187/packages/vitest/src/node/reporters/json.ts#L13-L63

type Milliseconds = number

interface Callsite {
  line: number
  column: number
}

export interface FormattedAssertionResult {
  ancestorTitles: Array<string>
  fullName: string
  status: 'passed' | 'failed' | 'skipped' | 'pending' | 'todo' | 'disabled'
  title: string
  duration?: Milliseconds | null
  failureMessages: Array<string>
  location?: Callsite | null
}

export interface FormattedTestResult {
  message: string
  name: string
  status: 'failed' | 'passed'
  startTime: number
  endTime: number
  assertionResults: Array<FormattedAssertionResult>
  // summary: string
  // coverage: unknown
}

export interface FormattedTestResults {
  numFailedTests: number
  numFailedTestSuites: number
  numPassedTests: number
  numPassedTestSuites: number
  numPendingTests: number
  numPendingTestSuites: number
  numTodoTests: number
  numTotalTests: number
  numTotalTestSuites: number
  startTime: number
  success: boolean
  testResults: Array<FormattedTestResult>
  // coverageMap?: CoverageMap | null | undefined
  // numRuntimeErrorTestSuites: number
  // snapshot: SnapshotSummary
  // wasInterrupted: boolean
}

export interface VitestInput {
  data: FormattedTestResults
  sha: string
  name?: string
}

export function vitestCheck({ data, sha, name = 'vitest' }: VitestInput): CheckRunCompleted {
  const result: CheckRunCompleted = {
    name,
    head_sha: sha,
    status: 'completed',
    conclusion: 'neutral',
    completed_at: new Date().toISOString(),
    output: {
      title: 'No tests found',
      summary: 'No tests found',
      annotations: []
    }
  }

  if (!data) {
    return result
  }

  result.conclusion = data.success ? 'success' : 'failure'

  if (!result.output) {
    return result
  }

  result.output.annotations = data.testResults
    .flatMap(results => {
      return results.assertionResults.map(assertionResult => {
        return {
          ...assertionResult,
          file: results.name
        }
      })
    })
    .filter(r => !['passed'].includes(r.status))
    .map<CheckAnnotation>(result => {
      const relPath = getRelativePath(result.file)

      let annotation_level: CheckAnnotationLevel
      switch (result.status) {
        case 'disabled':
          annotation_level = 'notice'
          break
        case 'failed':
          annotation_level = 'failure'
          break
        case 'pending':
          annotation_level = 'notice'
          break
        case 'skipped':
          annotation_level = 'notice'
          break
        case 'todo':
          annotation_level = 'notice'
          break
        default:
          annotation_level = 'notice'
      }
      const failureMessage = result.failureMessages?.join('\n')
      const message = result.status === 'pending' ? `skipped test '${result.title}'` : failureMessage

      let start_line = 1
      const lineColumnMatch = failureMessage.match(/sx?:(\d+):(\d+)\)/)
      if (lineColumnMatch) {
        start_line = parseInt(lineColumnMatch[1])
      }

      const annotation: CheckAnnotation = {
        start_line,
        end_line: start_line,
        annotation_level,
        title: [...result.ancestorTitles, result.title].join(' - '),
        message: message || 'unknown issue',
        path: relPath,
        raw_details: JSON.stringify(result, null, 2)
      }
      return annotation
    })
    // Limit to 50 annotations as this is the max per post for github
    .slice(0, 50)

  result.output.summary = `${data.numPassedTests} of ${data.numTotalTests} tests passed`
  // note: snapshots are not currently output in vitest reporter=json
  // note: check numTotalTestSuites for count?
  result.output.title = result.output.summary

  return result
}

function getRelativePath(path: string): string {
  const match = path.match(/^.*\/((?:src|bin)\/.+)$/)
  // Path needs to be set so default to a file we know exists
  return match && match.length === 2 ? match[1] : 'package.json'
}
