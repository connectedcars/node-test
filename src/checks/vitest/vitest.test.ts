import { vitestFailedOutput, vitestPassedOutput, vitestSkippedOutput } from './resources/vitest-help-text'
import { FormattedTestResults, vitestCheck } from './vitest'

describe('checks/vitest', () => {
  it('processes passing vitest output to checks structure', () => {
    const output = vitestCheck({ data: vitestPassedOutput, sha: '1234567890' })
    const expected = {
      name: 'vitest',
      head_sha: '1234567890',
      conclusion: 'success',
      status: 'completed',
      completed_at: expect.stringMatching(/^\d{4}/),
      output: {
        title: '116 of 116 tests passed',
        summary: '116 of 116 tests passed',
        annotations: []
      }
    }
    expect(output).toStrictEqual(expected)
  })

  it('processes failing vitest output to checks structure', () => {
    const output = vitestCheck({ data: vitestFailedOutput, sha: '1234567890' })
    expect(output).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })

  it('handles empty input', () => {
    const sampleOutput = ''
    const output = vitestCheck({
      data: sampleOutput as unknown as FormattedTestResults,
      sha: '1234567890'
    })
    expect(output).toStrictEqual({
      name: 'vitest',
      head_sha: '1234567890',
      conclusion: 'neutral',
      status: 'completed',
      completed_at: expect.stringMatching(/^\d{4}/),
      output: {
        title: 'No tests found',
        summary: 'No tests found',
        annotations: []
      }
    })
  })

  it('should handle skipped tests', () => {
    const output = vitestCheck({ data: vitestSkippedOutput, sha: '1234567890' })
    expect(output).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })
})
