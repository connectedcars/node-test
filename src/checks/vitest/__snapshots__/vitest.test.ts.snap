// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks/vitest processes failing vitest output to checks structure 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "failure",
  "head_sha": "1234567890",
  "name": "vitest",
  "output": Object {
    "annotations": Array [
      Object {
        "annotation_level": "failure",
        "end_line": 1,
        "message": "expected 'monday' to be 'tuesday' // Object.is equality",
        "path": "src/utils/day.test.ts",
        "raw_details": "{
  \\"ancestorTitles\\": [
    \\"\\",
    \\"day\\",
    \\"getDayFromDate\\"
  ],
  \\"fullName\\": \\" day getDayFromDate returns the correct day of week\\",
  \\"status\\": \\"failed\\",
  \\"title\\": \\"returns the correct day of week\\",
  \\"duration\\": 3,
  \\"failureMessages\\": [
    \\"expected 'monday' to be 'tuesday' // Object.is equality\\"
  ],
  \\"location\\": {
    \\"line\\": 17,
    \\"column\\": 43
  },
  \\"file\\": \\"/Users/mia/Developer/@connectedcars/manage/src/utils/day.test.ts\\"
}",
        "start_line": 1,
        "title": " - day - getDayFromDate - returns the correct day of week",
      },
    ],
    "summary": "115 of 116 tests passed",
    "title": "115 of 116 tests passed",
  },
  "status": "completed",
}
`;

exports[`checks/vitest should handle skipped tests 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "success",
  "head_sha": "1234567890",
  "name": "vitest",
  "output": Object {
    "annotations": Array [
      Object {
        "annotation_level": "notice",
        "end_line": 1,
        "message": "unknown issue",
        "path": "src/utils/graphql/query-generator.test.ts",
        "raw_details": "{
  \\"ancestorTitles\\": [
    \\"\\",
    \\"query-generator\\"
  ],
  \\"fullName\\": \\" query-generator parses columns to structure\\",
  \\"status\\": \\"skipped\\",
  \\"title\\": \\"parses columns to structure\\",
  \\"failureMessages\\": [],
  \\"file\\": \\"/Users/mia/Developer/@connectedcars/manage/src/utils/graphql/query-generator.test.ts\\"
}",
        "start_line": 1,
        "title": " - query-generator - parses columns to structure",
      },
      Object {
        "annotation_level": "notice",
        "end_line": 1,
        "message": "unknown issue",
        "path": "src/utils/graphql/query-generator.test.ts",
        "raw_details": "{
  \\"ancestorTitles\\": [
    \\"\\",
    \\"query-generator\\"
  ],
  \\"fullName\\": \\" query-generator is immutableish\\",
  \\"status\\": \\"skipped\\",
  \\"title\\": \\"is immutableish\\",
  \\"failureMessages\\": [],
  \\"file\\": \\"/Users/mia/Developer/@connectedcars/manage/src/utils/graphql/query-generator.test.ts\\"
}",
        "start_line": 1,
        "title": " - query-generator - is immutableish",
      },
      Object {
        "annotation_level": "notice",
        "end_line": 1,
        "message": "unknown issue",
        "path": "src/utils/graphql/query-generator.test.ts",
        "raw_details": "{
  \\"ancestorTitles\\": [
    \\"\\",
    \\"query-generator\\"
  ],
  \\"fullName\\": \\" query-generator adds variables to structure\\",
  \\"status\\": \\"skipped\\",
  \\"title\\": \\"adds variables to structure\\",
  \\"failureMessages\\": [],
  \\"file\\": \\"/Users/mia/Developer/@connectedcars/manage/src/utils/graphql/query-generator.test.ts\\"
}",
        "start_line": 1,
        "title": " - query-generator - adds variables to structure",
      },
      Object {
        "annotation_level": "notice",
        "end_line": 1,
        "message": "unknown issue",
        "path": "src/utils/graphql/query-generator.test.ts",
        "raw_details": "{
  \\"ancestorTitles\\": [
    \\"\\",
    \\"query-generator\\"
  ],
  \\"fullName\\": \\" query-generator correctly renders graphql union/interface types\\",
  \\"status\\": \\"skipped\\",
  \\"title\\": \\"correctly renders graphql union/interface types\\",
  \\"failureMessages\\": [],
  \\"file\\": \\"/Users/mia/Developer/@connectedcars/manage/src/utils/graphql/query-generator.test.ts\\"
}",
        "start_line": 1,
        "title": " - query-generator - correctly renders graphql union/interface types",
      },
    ],
    "summary": "116 of 116 tests passed",
    "title": "116 of 116 tests passed",
  },
  "status": "completed",
}
`;
